### PR TITLE
Blackjack

### PR DESCRIPTION
**Pull request by martmists**
Note: Do not merge until @aurieh adds the code for `CommandContext.waitFor`

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -2,7 +2,7 @@ group 'me.alexflipnote'
 version '1.0-SNAPSHOT'
 
 buildscript {
-    ext.kotlin_version = '1.2.21'
+    ext.kotlin_version = '1.2.31'
```
Comment (aurieh) [inline build.gradle]: any reason for this upgrade?

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
+package me.alexflipnote.kawaiibot.utils
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: better off in entities

File: src/main/kotlin/me/aurieh/ichigo/core/CommandContext.kt [hunk 3]
```diff
@@ -48,6 +49,14 @@ class CommandContext internal constructor(val receivedEvent: MessageReceivedEven
         }
     }
 
+    inline fun<reified T : Event> waitFor(crossinline predicate: (T) -> Boolean, timeout: Int = -1): T {
```
Comment (aurieh) [inline src/main/kotlin/me/aurieh/ichigo/core/CommandContext.kt]: should be suspend as well, though I will add it soon:tm:

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    private fun getPoints(cards: List<String>): Int {
+        var hasA = false
+        var points = 0
+        cards.forEach {
+           if (it[0] == 'A')
+               hasA = true
+            points += getValue(it)
+        }
+
+        if (points < 11 && hasA){
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: missing space

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                getPoints(userCards) == getPoints(botCards) -> {
+                    endStatus = "tied"
+                }
+                getPoints(userCards) < getPoints(botCards) -> {
+                    endStatus = "lost"
+                }
+            }
+        }
+
+        ctx.channel.sendMessage("${ctx.author.name} $endStatus against KawaiiBot!${getLayoutText(false)}")
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: wont send, #queue() or #await()

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    private suspend fun initialCards() {
+        botCards.add(drawCard())
+        userCards.add(drawCard())
+        botCards.add(drawCard())
+        userCards.add(drawCard())
+    }
+
+    private suspend fun askCard() {
+        var choice: String? = null
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: maybe use an empty string instead? nullability forces string to be a boxed type

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            } else {
+                deck.clear()  // Remove this line for actual blackjack instead of pure randomness (requested by alex)
+                userCards.clear()
+                botCards.clear()
+            }
+        }
+    }
+
+    private fun getLayoutText(hideBot: Boolean = true): String {
+        val botSet = "${botCards[0]} ${botCards.subList(1, botCards.size-1).map { it -> if (hideBot) "??" else it }.joinToString(" ")}"
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: lift this out of interpolation pls

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            if (getPoints(userCards) != 21) {
+                askCard()
+            }
+
+            val msg = ctx.channel.sendMessage("Do you want to play again? Type `yes` if you do.").await()
+            val choice = ctx.waitFor(ctx.sameContext(), 60).message.contentRaw
+            if (choice != "yes") {
+                playing = false
+            } else {
+                deck.clear()  // Remove this line for actual blackjack instead of pure randomness (requested by alex)
```
Comment (aurieh) [inline src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: @AlexFlipnote any comment on this?

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt (hunk 7 above)
Comment (aurieh) [reply src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: set a few locals

File: src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt (hunk 8 above)
Comment (AlexFlipnote) [reply src/main/kotlin/me/alexflipnote/kawaiibot/utils/BlackjackSession.kt]: I mentioned that I didn't want a predictable blackjack card deck, as the original was 100% random.

File: src/main/kotlin/me/aurieh/ichigo/core/CommandContext.kt (hunk 3 above)
Comment (martmists) [reply src/main/kotlin/me/aurieh/ichigo/core/CommandContext.kt]: Any progress on this yet?

**Comment by martmists** [comment]
My bad, misclicked

**Comment by Devoxin** [comment]
What's the status on this PR?

**Comment by martmists** [comment]
The only thing left is for Aurieh to fix [this](https://github.com/KawaiiBot/KawaiiBot/pull/7#discussion_r189688413), after which it should be able to get merged.

**Comment by Devoxin** [comment]
Should we just close this and iron out issues in the rewrite instead? There's been no progress on this PR since May 2018 and it doesn't look like those changes are going to be implemented any time soon -- equally I'm not sure if they are worth implementing as we have a rewrite branch that moves KawaiiBot over to a completely new command framework (although it might be worth mentioning there's little progress on that as we don't see the need to deploy any major updates yet.)

**Comment by AlexFlipnote** [comment]
Probably just close it, there haven't really been anyone asking "where's the blackjack command" at all since we removed it in Java rewrite.

**Pull request closed without merging**